src/charm.py

```diff
@@ -105,6 +108,24 @@ def _on_leader_elected(self, event) -> None:
 
         self._update_hosts(event)
 
+        if "db_initialised" not in self.app_data:
+            return
+
+        try:
+            self.update_app_relation_data()
+        except OperationFailure as e:
+            # when a unit departs the replica set it is possible that unit was a leader,
+            # thus resulting in mongod electing a new leader. If mongod hasn't elected a
+            # new leader yet an Authentication error will occur and we should retry updating
+            # the application relation data.
+            if e.code == 18:
+                logger.debug(
+                    "Waiting for replicaset to elect a new primary before updating relation data"
+                )
+                event.defer()
+            else:
+                raise
```

Comment:
Now that I think about it, it would be best to catch all PyMongoErrors. Sorry about that Abhishek.
```suggestion
        except PyMongoError as e:
            logger.error("Deferring on updating app relation data since: error: %r", e)
            event.defer()
            return
```